src/java/org/apache/cassandra/index/sai/disk/v3/BlockTerms.java

```diff
@@ -0,0 +1,1887 @@
+
+            this.termsIndexHandle = indexFiles.getFileAndCache(IndexComponent.BLOCK_TERMS_INDEX);
+            validateCRC(termsIndexHandle, meta.termsIndexCRC);
+
+            this.termsDataHandle = indexFiles.getFileAndCache(BLOCK_TERMS_DATA);
+            validateCRC(termsDataHandle, meta.termsCRC);
+
+            this.bitpackedHandle = indexFiles.getFileAndCache(IndexComponent.BLOCK_BITPACKED);
+            validateCRC(bitpackedHandle, meta.bitpackedCRC);
+
+            this.postingsHandle = indexFiles.getFileAndCache(IndexComponent.BLOCK_POSTINGS);
+            validateCRC(postingsHandle, meta.postingsCRC);
+
+            blockOrderMapFile = this.indexFiles.getFileAndCache(IndexComponent.BLOCK_ORDERMAP);
+            validateCRC(blockOrderMapFile, meta.orderMapCRC);
+
+            final SegmentMetadata.ComponentMetadata blockTermsOffsetsCompMeta = componentMetadatas.get(IndexComponent.BLOCK_TERMS_OFFSETS);
+            this.blockTermsOffsetsMeta = new NumericValuesMeta(blockTermsOffsetsCompMeta.attributes);
+
+            final MonotonicBlockPackedReader blockTermsOffsetsReader = new MonotonicBlockPackedReader(bitpackedHandle, blockTermsOffsetsMeta);
+            this.ramBytesUsed += blockTermsOffsetsReader.memoryUsage();
+            this.termBlockFPs = blockTermsOffsetsReader.open();
+
+            final SegmentMetadata.ComponentMetadata blockPostingsOffsetsCompMeta = componentMetadatas.get(IndexComponent.BLOCK_POSTINGS_OFFSETS);
+            this.blockPostingsOffsetsMeta = new NumericValuesMeta(blockPostingsOffsetsCompMeta.attributes);
+            final MonotonicBlockPackedReader blockPostingOffsetsReader = new MonotonicBlockPackedReader(bitpackedHandle, blockPostingsOffsetsMeta);
+            this.ramBytesUsed += blockPostingOffsetsReader.memoryUsage();
+            this.postingBlockFPs = blockPostingOffsetsReader.open();
+
+            final SegmentMetadata.ComponentMetadata blockOrderMapCompMeta = componentMetadatas.get(IndexComponent.BLOCK_ORDERMAP);
+            this.blockOrderMapMeta = new NumericValuesMeta(blockOrderMapCompMeta.attributes);
+            final BlockPackedReader blockOrderMapFPReader = new BlockPackedReader(bitpackedHandle, blockOrderMapMeta);
+            this.ramBytesUsed += blockOrderMapFPReader.memoryUsage();
+            this.orderMapFPs = blockOrderMapFPReader.open();
+
+            if (meta.numPostingBlocks != postingBlockFPs.length())
+                throw new IllegalStateException();
+
+            if (meta.numPostingBlocks != orderMapFPs.length())
+                throw new IllegalStateException();
+
+            this.orderMapReader = DirectReaders.getReaderForBitsPerValue((byte) DirectWriter.unsignedBitsRequired(meta.postingsBlockSize - 1));
+            this.orderMapRandoInput = new SeekingRandomAccessInput(IndexInputReader.create(blockOrderMapFile));
+
+            if (indexDescriptor.componentExists(IndexComponent.BLOCK_UPPER_POSTINGS_OFFSETS, context, indexFiles.isTemporary()))
```

Comment:
Under what circumstances will the `BLOCK_UPPER_POSTINGS_OFFSETS` file not exist? It seems that the only other usage of `upperPostingsReader` is in `search` and that will NPE if this is not set.